### PR TITLE
Add SharedPool

### PR DESCRIPTION
**Pull request by friedmud**
Ready to go.

refs #12030

### CODE REVIEW

File: framework/include/utils/SharedPool.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+// System Includes
+#include <stack>
+#include <memory>
+
+namespace MooseUtils
+{
+
+template <class T, typename... Args>
+auto
+reset(int, T & obj, Args... args) -> decltype(obj.reset(args...), void())
```
Comment (permcody) [inline framework/include/utils/SharedPool.h]: I believe this requires C++14 syntax. Adding the minimum targets to see what happens on our build systems.  Can you comment on why this is needed? If it forces a compiler upgrade, we'll want to know the pros/cons.

Comment (friedmud) [reply framework/include/utils/SharedPool.h]: It's needed to be able to detect whether a `reset()` function exists so we can call it when acquiring an existing object from the pool.

As you noted: this should be legal C++11

**Comment by dschwen** [comment]
This may be a super dumb question, but how is a shared pool different from a heap? Is there lower overhead, an advantage from administering objects that always have the same size(?)...

**Review by permcody — APPROVED** [review]

**Comment by friedmud** [comment]
@dschwen in general: malloc is slow.  It's a syscall to the OS and has a long round-trip time... in addition it's typically a critical section so all threads block.

Also: lots of allocation/deallocation can cause memory fragmentation... which makes it take even longer to get a decently sized block of memory.  It can also mean that it can "run out" of memory.

The Pool helps in many ways... but the one I'll highlight is that it holds whole _objects_.  Those objects can have many vectors in them that are already sized... so it can remove the need for _many_ mallocs (not just one).  If you're creating and destroying MANY objects like this (like I am) it can be a large speed boost.

**Comment by dschwen** [comment]
And there will be a pool for each thread?

**Comment by friedmud** [comment]
@dschwen You could do that - or make the sharedpool have some locks in it.  Right now I just don't use it in threaded regions (but that will be changing soon - and my plan is to put an optional template parameter that turns on locking)

**Pull request merged**